pkg/component/component.go

```diff
@@ -0,0 +1,85 @@
+package component
+
+import (
+	"strings"
+
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+)
+
+// Component that implements gRPC StoreApi
+type StoreAPI interface {
+	implementsStoreAPI()
+	String() string
+	ToProto() storepb.StoreType
+}
+
+// Component that produce blocks of metrics
+type Source interface {
+	producesBlocks()
+	String() string
+}
+
+// Component that implements gRPC StoreApi and produce blocks of metrics
+type SourceStoreAPI interface {
+	implementsStoreAPI()
+	producesBlocks()
```

Comment:
Again, we could use embedding here. For example: 

```
type SourceStoreAPI interface {
    Source
    StoreAPI
}
```

Reply:
But I guess it will cause diamond problem as described here: https://joaodlf.com/go-the-diamond-problem.html 

(with `String()`) - that's fine then - let's leave your version as it is.